File: src/main/java/com/starshootercity/abilities/MasterOfWebs.java
package com.starshootercity.abilities;

import com.destroystokyo.paper.event.server.ServerTickEndEvent;
import com.starshootercity.OriginSwapper;
import com.starshootercity.OriginsReborn;
import net.kyori.adventure.key.Key;
import net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket;
import net.minecraft.network.syncher.EntityDataAccessor;
import net.minecraft.network.syncher.EntityDataSerializers;
import net.minecraft.network.syncher.SynchedEntityData;
import net.minecraft.server.level.ServerPlayer;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.craftbukkit.v1_20_R3.entity.CraftEntity;
import org.bukkit.craftbukkit.v1_20_R3.entity.CraftPlayer;
import org.bukkit.entity.Entity;
import org.bukkit.entity.HumanEntity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockDropItemEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.inventory.PrepareItemCraftEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.ShapelessRecipe;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MasterOfWebs implements FlightAllowingAbility, Listener, VisibleAbility {
    private final Map<Player, List<Entity>> glowingEntities = new HashMap<>();

    private final Map<Player, Integer> cobwebSpawnCooldowns = new HashMap<>();
    private final List<Location> temporaryCobwebs = new ArrayList<>();

    @EventHandler
    public void onBlockDropItem(BlockDropItemEvent event) {
        if (temporaryCobwebs.contains(event.getBlock().getLocation())) {
            event.setCancelled(true);
            temporaryCobwebs.remove(event.getBlock().getLocation());
        }
    }

    @EventHandler
    public void onEntityDamageByEntity(EntityDamageByEntityEvent event) {
        if (event.getDamager() instanceof Player player) {
            AbilityRegister.runForAbility(player, getKey(), () -> {
                Integer cooldown = cobwebSpawnCooldowns.get(player);
                if ((cooldown == null || Bukkit.getCurrentTick() - cooldown >= 120) && !event.getEntity().getLocation().getBlock().isSolid()) {
                    cobwebSpawnCooldowns.put(player, Bukkit.getCurrentTick());
                    Location location = event.getEntity().getLocation().getBlock().getLocation();
                    temporaryCobwebs.add(location);
                    location.getBlock().setType(Material.COBWEB);
                    Bukkit.getScheduler().scheduleSyncDelayedTask(OriginsReborn.getInstance(), () -> {
                        if (location.getBlock().getType() == Material.COBWEB && temporaryCobwebs.contains(location)) {
                            temporaryCobwebs.remove(location);
                            location.getBlock().setType(Material.AIR);
                        }
                    }, 60);
                }
            });
        }
    }


    private void setCanFly(Player player, boolean setFly) {
        if (setFly) player.setAllowFlight(true);
        canFly.put(player, setFly);
    }

    private final Map<Player, Boolean> canFly = new HashMap<>();


    @EventHandler
    public void onServerTickEnd(ServerTickEndEvent event) {
        for (Player webMaster : Bukkit.getOnlinePlayers()) {
            AbilityRegister.runForAbility(webMaster, getKey(), () -> {
                if (isInCobweb(webMaster)) {
                    setCanFly(webMaster, true);
                    webMaster.setFlying(true);
                } else {
                    setCanFly(webMaster, false);
                }
                List<Entity> entities = webMaster.getNearbyEntities(16, 16, 16);
                entities.removeIf(entity -> !(entity instanceof LivingEntity));
                if (entities.size() > 16) entities = entities.subList(0, 16);
                entities.addAll(Bukkit.getOnlinePlayers());
                entities.removeIf(entity -> entity.getWorld() != webMaster.getWorld());
                entities.removeIf(entity -> entity.getLocation().distance(webMaster.getLocation()) > 16);
                for (Entity webStuck : entities) {
                    AbilityRegister.runWithoutAbility(webStuck, getKey(), () -> {
                        if (webStuck != webMaster) {
                            if (!glowingEntities.containsKey(webMaster)) {
                                glowingEntities.put(webMaster, new ArrayList<>());
                            }
                            if (isInCobweb(webStuck)) {
                                if (!glowingEntities.get(webMaster).contains(webStuck)) {
                                    glowingEntities.get(webMaster).add(webStuck);
                                }

                                ServerPlayer master = ((CraftPlayer) webMaster).getHandle();
                                net.minecraft.world.entity.Entity stuck = ((CraftEntity) webStuck).getHandle();

                                byte data = 0x40;
                                if (webStuck.getFireTicks() > 0) {
                                    data += 0x01;
                                }
                                if (webStuck.isSneaking()) {
                                    data += 0x02;
                                }
                                if (webStuck instanceof Player stuckPlayer) {
                                    if (stuckPlayer.isSprinting()) {
                                        data += 0x08;
                                    }
                                    if (stuckPlayer.isSwimming()) {
                                        data += 0x10;
                                    }
                                    if (stuckPlayer.isInvisible()) {
                                        data += 0x20;
                                    }
                                    if (stuckPlayer.isGliding()) {
                                        data += 0x80;
                                    }
                                }
                                List<SynchedEntityData.DataValue<?>> eData = new ArrayList<>();
                                eData.add(SynchedEntityData.DataValue.create(new EntityDataAccessor<>(0, EntityDataSerializers.BYTE), data));
                                ClientboundSetEntityDataPacket metadata = new ClientboundSetEntityDataPacket(stuck.getId(), eData);
                                master.connection.send(metadata);
                            } else {
                                glowingEntities.get(webMaster).remove(webStuck);
                                AbilityRegister.updateEntity(webMaster, webStuck);
                            }
                        }
                    });
                }
            });
        }
    }

    public MasterOfWebs() {
        NamespacedKey recipeKey = new NamespacedKey(OriginsReborn.getInstance(), "web-recipe");
        ShapelessRecipe webRecipe = new ShapelessRecipe(recipeKey, new ItemStack(Material.COBWEB));
        if (Bukkit.getRecipe(recipeKey) == null) {
            webRecipe.addIngredient(Material.STRING);
            webRecipe.addIngredient(Material.STRING);
            Bukkit.addRecipe(webRecipe);
        }
    }

    @EventHandler
    public void onPrepareItemCraft(PrepareItemCraftEvent event) {
        if (event.getRecipe() != null) {
            if (event.getRecipe().getResult().getType() == Material.COBWEB) {
                for (HumanEntity entity : event.getInventory().getViewers()) {
                    if (entity instanceof Player player) {
                        AbilityRegister.runWithoutAbility(player, getKey(), () -> event.getInventory().setResult(null));
                    }
                }
            }
        }
    }

    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:master_of_webs");
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getDescription() {
        return OriginSwapper.LineData.makeLineFor("You navigate cobweb perfectly, and are able to climb in them. When you hit an enemy in melee, they get stuck in cobweb for a while. Non-arthropods stuck in cobweb will be sensed by you. You are able to craft cobweb from string.", OriginSwapper.LineData.LineComponent.LineType.DESCRIPTION);
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getTitle() {
        return OriginSwapper.LineData.makeLineFor("Master of Webs", OriginSwapper.LineData.LineComponent.LineType.TITLE);
    }

    public boolean isInCobweb(Entity entity) {
        for (Block start : new ArrayList<Block>() {{
            add(entity.getLocation().getBlock());
            add(entity.getLocation().getBlock().getRelative(BlockFace.UP));
        }}) {
            if (start.getType() == Material.COBWEB) return true;
            for (BlockFace face : BlockFace.values()) {
                Block block = start.getRelative(face);
                if (block.getType() != Material.COBWEB) continue;
                if (entity.getBoundingBox().overlaps(block.getBoundingBox())) {
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public boolean canFly(Player player) {
        return canFly.getOrDefault(player, false);
    }

    @Override
    public float getFlightSpeed(Player player) {
        return 0.04f;
    }
}

File: core/src/main/java/com/starshootercity/abilities/MasterOfWebs.kt
package com.starshootercity.abilities

import com.destroystokyo.paper.event.server.ServerTickEndEvent
import com.starshootercity.OriginSwapper.LineData.Companion.makeLineFor
import com.starshootercity.OriginSwapper.LineData.LineComponent
import com.starshootercity.OriginsReborn.Companion.NMSInvoker
import com.starshootercity.OriginsReborn.Companion.instance
import com.starshootercity.abilities.Ability.AbilityRunner
import com.starshootercity.cooldowns.CooldownAbility
import com.starshootercity.cooldowns.Cooldowns.CooldownInfo
import net.kyori.adventure.key.Key
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.block.Block
import org.bukkit.block.BlockFace
import org.bukkit.entity.Entity
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockDropItemEvent
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.inventory.PrepareItemCraftEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.ShapelessRecipe

class MasterOfWebs : CooldownAbility, FlightAllowingAbility, Listener, VisibleAbility {
    private val glowingEntities: MutableMap<Player?, MutableList<Entity?>?> = HashMap<Player?, MutableList<Entity?>?>()

    private val temporaryCobwebs: MutableList<Location?> = ArrayList<Location?>()

    @EventHandler
    fun onBlockDropItem(event: BlockDropItemEvent) {
        if (temporaryCobwebs.contains(event.getBlock().location)) {
            event.isCancelled = true
            temporaryCobwebs.remove(event.getBlock().location)
        }
    }

    @EventHandler
    fun onEntityDamageByEntity(event: EntityDamageByEntityEvent) {
        runForAbility(event.damager, AbilityRunner { player: Player? ->
            if (hasCooldown(player!!)) return@AbilityRunner
            if (!event.getEntity().location.block.isSolid) {
                setCooldown(player)
                val location = event.getEntity().location.block.location
                temporaryCobwebs.add(location)
                location.block.type = Material.COBWEB
                Bukkit.getScheduler().scheduleSyncDelayedTask(instance, Runnable {
                    if (location.block.type == Material.COBWEB && temporaryCobwebs.contains(location)) {
                        temporaryCobwebs.remove(location)
                        location.block.type = Material.AIR
                    }
                }, 60)
            }
        })
    }


    private fun setCanFly(player: Player, setFly: Boolean) {
        if (setFly) player.allowFlight = true
        canFly.put(player, setFly)
    }

    private val canFly: MutableMap<Player?, Boolean?> = HashMap<Player?, Boolean?>()


    @EventHandler
    fun onServerTickEnd(event: ServerTickEndEvent?) {
        val onlinePlayers = Bukkit.getOnlinePlayers()

        for (player in onlinePlayers) {
            runForAbility(player, AbilityRunner { webMaster: Player ->
                val inCobweb = isInCobweb(webMaster)
                setCanFly(webMaster, inCobweb)
                if (inCobweb) webMaster.isFlying = true

                val entities = (webMaster.getNearbyEntities(16.0, 16.0, 16.0)
                    .filterIsInstance<LivingEntity>()
                    .take(16) + onlinePlayers)
                    .distinct()
                    .filter { it.world == webMaster.world && it.location.distance(webMaster.location) <= 16 }

                for (entity in entities) {
                    runForAbility(entity, null) { webStuck: Player ->
                        if (webStuck !== webMaster) {
                            val masterEntities = glowingEntities.getOrPut(webMaster) { ArrayList() }!!
                            if (isInCobweb(webStuck)) {
                                if (!masterEntities.contains(webStuck)) {
                                    masterEntities.add(webStuck)
                                }
                                NMSInvoker.sendEntityData(webMaster, webStuck, getData(webStuck))
                            } else {
                                masterEntities.remove(webStuck)
                                AbilityRegister.updateEntity(webMaster, webStuck)
                            }
                        }
                    }
                }
            })
        }
    }


    init {
        val recipeKey = NamespacedKey(instance, "web-recipe")
        val webRecipe = ShapelessRecipe(recipeKey, ItemStack(Material.COBWEB))
        if (Bukkit.getRecipe(recipeKey) == null) {
            webRecipe.addIngredient(Material.STRING)
            webRecipe.addIngredient(Material.STRING)
            Bukkit.addRecipe(webRecipe)
        }
    }

    @EventHandler
    fun onPrepareItemCraft(event: PrepareItemCraftEvent) {
        if (event.recipe != null) {
            if (event.recipe!!.result.type == Material.COBWEB) {
                for (entity in event.inventory.viewers) {
                    runForAbility(
                        entity,
                        null,
                        AbilityRunner { player: Player? -> event.inventory.result = null })
                }
            }
        }
    }

    override fun getKey(): Key {
        return Key.key("origins:master_of_webs")
    }

    override val description: MutableList<LineComponent> = makeLineFor(
        "You navigate cobweb perfectly, and are able to climb in them. When you hit an enemy in melee, they get stuck in cobweb for a while. Non-arthropods stuck in cobweb will be sensed by you. You are able to craft cobweb from string.",
        LineComponent.LineType.DESCRIPTION
    )

    override val title: MutableList<LineComponent> = makeLineFor(
        "Master of Webs",
        LineComponent.LineType.TITLE
    )

    fun isInCobweb(entity: Entity): Boolean {
        for (start in object : ArrayList<Block?>() {
            init {
                add(entity.location.block)
                add(entity.location.block.getRelative(BlockFace.UP))
            }
        }) {
            if (start!!.type == Material.COBWEB) return true
            for (face in BlockFace.entries) {
                val block = start.getRelative(face)
                if (block.type != Material.COBWEB) continue
                if (entity.boundingBox.overlaps(block.boundingBox)) {
                    return true
                }
            }
        }
        return false
    }

    override fun canFly(player: Player): Boolean {
        return canFly.getOrDefault(player, false)!!
    }

    override fun getFlightSpeed(player: Player): Float {
        return 0.04f
    }

    override val cooldownInfo: CooldownInfo
        get() = CooldownInfo(120, "web")

    companion object {
        private fun getData(webStuck: Entity): Byte {
            var data: Byte = 0x40
            if (webStuck.fireTicks > 0) {
                data = (data + 0x01).toByte()
            }
            if (webStuck is LivingEntity) {
                if (webStuck.isInvisible) data = (data + 0x20).toByte()
            }
            if (webStuck is Player) {
                if (webStuck.isSneaking) {
                    data = (data + 0x02).toByte()
                }
                if (webStuck.isSprinting) {
                    data = (data + 0x08).toByte()
                }
                if (webStuck.isSwimming) {
                    data = (data + 0x10).toByte()
                }
                if (webStuck.isGliding) {
                    data = (data + 0x80.toByte()).toByte()
                }
            }
            return data
        }
    }
}
